### ISSUE TITLE
Issues with SQL Server Spatial function STAsText() 

### ISSUE THREAD

**Issue by cmhh** [rstats-db/odbc#82]
I guess this won't be a particularly mainstream issue, but I have the need to extract data from SQL Server Spatial databases on Linux, and using the official Microsoft ODBC driver.  This driver cannot be used to extract geography types without truncating them (the Windows driver works fine), but I was hopeful of being able to extract geographies as WKT using the STAsText() function.  When I do this, however, I generally get an error.  As an example:

```r
library(DBI)
library(odbc)

con <- dbConnect(odbc::odbc(), 
                 Driver = "SQL Server",
                 Server = "192.168.20.21",
                 Port = 1433,
                 Database  = "gis",
                 Schema = "geography",
                 UID = "gisuser",
                 PWD = "password")

shp <- dbGetQuery(con, "select f_geometry_column from dbo.geometry_columns where f_table_name = 'mb2013'")[1,1]
cols <- dbListFields(con, "mb2013")
query <- sprintf("select top 10 %s, %s.STAsText() as WKT from geography.mb2013",
                 paste(cols, collapse = ", "),
                 shp)

res <- dbGetQuery(con, query)
```

produces the error:

```as.is
Error in result_fetch(res@ptr, n, ...) : 
  nanodbc/nanodbc.cpp:2834: 07009: [Microsoft][ODBC Driver 13 for S
```

Curiously, it sometimes works.  For example, the following query works:

```as.is
select top 10 ogr_fid, mb2013, ogr_geometry.STAsText() WKT from geography.mb2013
```

but this doesn't (same query, but columns in a different order):

```as.is
select top 10 ogr_fid, ogr_geometry.STAsText() as WKT, mb2013 from geography.mb2013
```

It produces a slightly different error, though (different line number):

```as.is
Error in result_fetch(res@ptr, n, ...) : 
  nanodbc/nanodbc.cpp:2890: 07009: [Microsoft][ODBC Driver 13 for S
```

I get the similar errors on Windows, but again with different numbers `nanodbc/nanodbc.cpp:xxxx`.

Note that this all works if I switch to using the `RJDBC` package with the official MS JDBC driver, but for various reasons I'm trying to avoid a solution which uses the JVM... plus I just really like this `odbc` package and how easily I can use it with `dbplyr`, for example.

Anyway, I appreciate that this might be hard to reproduce, or reproducing it might take a bit of work, but thought I'd ask the question anyway.

**Comment by cmhh** [comment]
As an additional point, the geometry fields converted to WKT in this situation can be very long--several hundred thousand characters, for example.

**Comment by cmhh** [comment]
Oh, the error from Windows includes a little more text...

```as.is
Error in result_fetch(res@ptr, n, ...) : 
  nanodbc/nanodbc.cpp:3110: 07009: [Microsoft][ODBC SQL Server Driver]Invalid Descriptor Index
```

**Comment by jimhester** [comment]
Please verify you are using the latest CRAN version of ODBC.

**Comment by jcfisher** [comment]
I am no expert, but "Invalid descriptor index" occurring when columns are selected out of order seems to be a general problem with MS SQL Server.  See [this Stack Overflow answer](https://stackoverflow.com/a/6361520/2520640).

**Comment by jimhester** [comment]
Yes you are right, it is a known issue with the Microsoft ODBC driver for SQL Server (https://github.com/rstats-db/odbc/issues/10). The workaround is to always put `varchar(max)` columns at the end of the select query.

**Comment by cmhh** [comment]
Yes, I tried the version on CRAN, and also with `devtools::install_github`.

Interesting about the column order thing. All news to me--not a huge SQL Server user, but it is the main RDBMS in use where I work so I have little choice.  I also had the `odbc` package throw an error today when selecting a column with type `nvarchar(max)`. I had to use the convert function to work around this and set the type to `varchar(255)`.

I hadn't really had any trouble with the `odbc` package until I tried using it with these spatial tables, but I'll look a bit harder at the linked stackoverflow thread and see if I can figure something out.  But the JDBC driver works in a pinch--as long as `rJava` is configured okay, the heap size is increased, and the JVM can be started (we have a weird issue at work where this often fails because the per user thread count has been exceeded due to lots of zombie r-session processes).

Thanks.
